### ISSUE TITLE
Android: Native Compose Library

### ISSUE THREAD

**Issue by Doomsdayrs** [servo/servo#33742]
**Describe the new feature:**

Currently, The Android support directory contains an example application and a view library for Servo. 

The main problem I see is that not only are they written using Java, but also using the Android native view toolkits instead of Compose.

Guidance for all new Android applications is to be written in Kotlin and using Compose.

There are a few ways this can be accomplished as I see it.

1. Conversion of the android sub-directory completely to Kotlin/Compose.
2. Creation of a Kotlin/Compose sub-module that exists separately of the current code base.
3. Extraction of Servo native interop code into a separate sub-module and perform option two.

I plan in the coming weeks to slowly send patches at least converting the current gradle files to Kotlin in order to better facilitate a migration (or no migration, I will send anyways since groovy is a mess).

**Comment by msub2** [comment]
Pinging @mukilan for their thoughts

**Comment by mukilan** [comment]
Moving to Kotlin is a great idea! This is something we wanted to do at some point. I'm not too familiar with Compose, so I don't have a strong opinion on that.  I guess it should be fine as long as we can still distribute `ServoView` in a way that third-party applications can embed it in their own apps and we can potentially extend it into a [WebView provider](https://chromium.googlesource.com/chromium/src/+/HEAD/android_webview/docs/webview-providers.md) in the future.

**Comment by wusyong** [comment]
There's [GeckoView](https://mozilla.github.io/geckoview/) we can reference from.

**Comment by sagudev** [comment]
I always wanted to see engine-servo as mozilla android component: https://github.com/mozilla-mobile/firefox-android/tree/main/android-components/components/browser/engine-system.

**Comment by Doomsdayrs** [comment]
I feel like discussion of Servo as an android component for those systems is out of scope of this issue, but a separate issue for it would be wise.

**Comment by wusyong** [comment]
@Doomsdayrs I think it's somewhat related a bit. Could this library evolve into something like geckoview in the future? Or do we have to rewrite everything all over again.

**Comment by Doomsdayrs** [comment]
> @Doomsdayrs I think it's somewhat related a bit. Could this library evolve into something like geckoview in the future? Or do we have to rewrite everything all over again.

I do not see a reason why this library would be unable to be adapted to fit into such a schema. This statement is made under my current understanding that: `GeckoView` is simply a UI component that is added to an application, `engine-servo` would be an export of the business code for integration into... Firefox(?), and `webview provider` would also just be an export of business code but for system wide integration(?). If anyone more educated on those matters finds any objection, feel free to correct me.

The principal of this issue is to utilize modern tools (Kotlin / Compose) to build up the library, which enables more fluid integration with new android applications, design language, and schemas. 

Google ported their [Maps library](https://github.com/googlemaps/android-maps-compose) to compose [as well](https://developers.google.com/maps/documentation/android-sdk/maps-compose), which may serve as an example for servo of how to build up a library around Android compose.

For legacy applications, they can [embed](https://developer.android.com/develop/ui/compose/migrate/interoperability-apis/compose-in-views) compose into android views, which of course they will maintain for their legacy systems. But I doubt there are many legacy apps gearing up to integrate Servo, and new apps should be using compose by default.

**Comment by Doomsdayrs** [comment]
Working on figuring out the threading that Servo uses for proper migration to compose and coroutines.

@paulrouget 

What was your rationale behind `inGLThread` vs `inUIThread`?
What even is the "GL" thread?

There is no documentation in the code base.

**Comment by mukilan** [comment]
@Doomsdayrs: Android requires that any OpenGL rendering [happen in a thread that is not the UI thread](https://source.android.com/docs/core/graphics/arch-sv-glsv#activity), so we need to run the servo instance (the compositor thread specifically) its own thread and pass any events from Android to this other thread. `inGLThread` and `inUIThread` are helpers to achieve this message passing between the threads using the [`Handler.post`](https://developer.android.com/reference/android/os/Handler#post(java.lang.Runnable)) method.

In the original Android code, the `ServoView` class that represents the an instance of Servo was derived from the `GLSurfaceView` base class. Deriving from this class means Android will take care the threading aspect (and GL thread referred to this thread). We had to switch to the `SurfaceView` base class when we revived the Android code, so now the GL thread is just the thread we create [here](https://github.com/servo/servo/blob/fc569260d201dd947aff1d9aa2e824c95fa26f2a/support/android/apk/servoview/src/main/java/org/servo/servoview/ServoView.java#L88).

**Comment by Doomsdayrs** [comment]
Thank you for better explaining the design choices.

---

> We had to switch to the `SurfaceView`

What was the reason behind the switch?

**Comment by Doomsdayrs** [comment]
In `Servo.shutdown()`, the following code can be found (excerpt)

```java
// Wait until Servo gets back to us to finalize shutdown.
while (!mShutdownComplete) {
 ...
 Thread.sleep(10);
 ...
 mJNI.performUpdates();
}
```

What is the purpose of `mJNI.performUpdates()` here?

**Comment by veyndan** [comment]
As one can see from the linked issues, it is probably unsurprising that I'd like to resurrect this issue 😄 

To summarise the updates:

- All Android code has been converted to Kotlin (#45640, #46701).
- The app module has been converted to Compose UI (#45715). Technically independent of this issue, but can be considered to show a precedent.
- The conversion of the library module to Kotlin in particular was always a prerequisite to getting this issue off the ground, so this issue can now be considered unblocked.

I don't want to restate anything that's already mentioned. The main qualms seem to be:

1. Wanting to provide something like `GeckoView`/`engine-servo` (which I agree with @Doomsdayrs that [this seems like a separate issue](https://github.com/servo/servo/issues/33742#issuecomment-2408285578)).
2. Being able to use it in non-Compose UI code, which as @Doomsdayrs [already mentioned would be possible](https://github.com/servo/servo/issues/33742#issuecomment-2408324744).
3. In more recent news (quoting https://github.com/servo/servo/issues/45715), the relevancy of Android Views have been waning with time and has been [officially moved into maintenance mode, with Compose UI being the actively developed, recommended Android UI toolkit](https://android-developers.googleblog.com/2026/05/android-ui-development-is-compose-first.html). This just further solidifies what was said at https://github.com/servo/servo/issues/33742#issuecomment-2408324744, where "[there aren't] many legacy apps gearing up to integrate Servo, and new apps should be using compose by default."

Assuming there are no objections (and no one is currently working on it), I'd like to begin work on migrating this module over to Compose UI! I have no intention of steamrolling this in, so I'll wait a few days before beginning work on this in case of concerns.
